src/adapter/src/client.rs

```diff
@@ -394,147 +521,45 @@ impl SessionClient {
     /// interface. The provided `stmts` are executed directly, and their results
     /// are returned as a vector of rows, where each row is a vector of JSON
     /// objects.
+    ///
+    /// n.b. To handle committing statements, this process requires breaking
+    /// after generating a `SimpleResult::err`.
     pub async fn simple_execute(
         &mut self,
         stmts: &str,
     ) -> Result<SimpleExecuteResponse, AdapterError> {
         let stmts =
             mz_sql::parse::parse(&stmts).map_err(|e| AdapterError::Unstructured(e.into()))?;
+
         let num_stmts = stmts.len();
-        const EMPTY_PORTAL: &str = "";
         let mut results = vec![];
-        for stmt in stmts {
-            // Mirror the behavior of the PostgreSQL simple query protocol.
-            // See the pgwire::protocol::StateMachine::query method for details.
-            self.start_transaction(Some(num_stmts)).await?;
 
-            if let Err(e) = self.declare(EMPTY_PORTAL.into(), stmt, vec![]).await {
-                results.push(SimpleResult::err(e));
-                continue;
-            }
-
-            let desc = self
-                .session()
-                // We do not need to verify here because `self.execute` verifies below.
-                .get_portal_unverified(EMPTY_PORTAL)
-                .map(|portal| portal.desc.clone())
-                .expect("unnamed portal should be present");
-            if !desc.param_types.is_empty() {
-                results.push(SimpleResult::err("query parameters are not supported"));
-                continue;
+        for stmt in stmts {
+            if matches!(self.session().transaction(), TransactionStatus::Failed(_)) {
```

Comment:
This needs to allow certain transaction exit statements, should be able to copy https://github.com/MaterializeInc/materialize/blob/608aacb7c8154afb767e065bd269dac191c4eefc/src/pgwire/src/protocol.rs#L468

Reply:
Doing this with the current machinery allows statements after the failed transaction to execute, which sounded like it was prohibited with the simple protocol. Can you suggest either a means to test that behavior in PG, i.e. is pgtest the extended or simple protocol, or a fix for this that you would be comfortable with?

Reply:
This is specifically allowed by postgres:

> However, if the Query message contains a COMMIT or ROLLBACK closing the existing transaction block, then any following statements are executed in an implicit transaction block.

https://github.com/postgres/postgres/blob/4b4663fb4a704b7a145ed6ec018b1f0c116eeb6b/src/backend/tcop/postgres.c#L1088

https://www.postgresql.org/docs/current/protocol-flow.html#id-1.10.5.7.4

Reply:
From https://www.postgresql.org/docs/current/protocol-flow.html#id-1.10.5.7.4

Remember that, regardless of any transaction control commands that may be present, execution of the Query message stops at the first error. Thus for example given

```sql
BEGIN;
SELECT 1/0;
ROLLBACK;
```

in a single Query message, the session will be left inside a failed regular transaction block, since the ROLLBACK is not reached after the divide-by-zero error. Another ROLLBACK will be needed to restore the session to a usable state.

Reply:
Ahh that postgres code is checking if it's in an explicit aborted transaction? An implicit aborted transaction has perhaps already left the loop? Unclear, but confirmed you're correct here. https://github.com/MaterializeInc/materialize/pull/14698

Reply:
Are you all right with the submitted approach for the simple protocol in light of that? This will get rebased over the PR that adds the extended protocol, where we will use the behavior you've pointed out in the earlier comments.

Reply:
Yes, this can go ahead. I'm also positive no user cares about these details, and we can fix if we discover edge cases.